File: crates/brainwires-mdap/examples/voting_consensus.rs
//! Voting Consensus Example
//!
//! Demonstrates the building blocks of the MAKER voting system:
//! creating sampled responses with metadata, configuring red-flag
//! validation, and validating responses against reliability criteria.
//!
//! Run with: `cargo run -p brainwires-agent --features mdap --example voting_consensus`

use brainwires_mdap::red_flags::{
    OutputFormat, RedFlagConfig, RedFlagValidator, StandardRedFlagValidator,
};
use brainwires_mdap::voting::{ResponseMetadata, SampledResponse, VoteResult, VotingMethod};
use std::collections::HashMap;

fn main() {
    println!("=== MAKER Voting Consensus Building Blocks ===\n");

    // 1. Create sampled responses with metadata
    println!("--- 1. Creating Sampled Responses ---\n");

    let responses = [
        SampledResponse::new(
            "Paris".to_string(),
            ResponseMetadata {
                token_count: 12,
                response_time_ms: 150,
                format_valid: true,
                finish_reason: Some("stop".to_string()),
                model: Some("claude-sonnet".to_string()),
            },
            "The capital of France is Paris.".to_string(),
        ),
        SampledResponse::with_confidence(
            "Paris".to_string(),
            ResponseMetadata {
                token_count: 8,
                response_time_ms: 120,
                format_valid: true,
                finish_reason: Some("stop".to_string()),
                model: Some("claude-sonnet".to_string()),
            },
            "Paris".to_string(),
            0.95,
        ),
        SampledResponse::new(
            "Wait, actually Lyon".to_string(),
            ResponseMetadata {
                token_count: 45,
                response_time_ms: 300,
                format_valid: true,
                finish_reason: Some("stop".to_string()),
                model: None,
            },
            "Wait, I think it might be Lyon. Actually, let me reconsider...".to_string(),
        ),
        SampledResponse::new(
            "Paris".to_string(),
            ResponseMetadata {
                token_count: 900,
                response_time_ms: 2000,
                format_valid: true,
                finish_reason: Some("length".to_string()),
                model: Some("gpt-4o".to_string()),
            },
            "A very long response that was truncated...".to_string(),
        ),
    ];

    for (i, resp) in responses.iter().enumerate() {
        println!(
            "  Response {}: value={:?}, tokens={}, confidence={:.2}, model={:?}",
            i + 1,
            resp.value,
            resp.metadata.token_count,
            resp.confidence,
            resp.metadata.model,
        );
    }

    // 2. Compare RedFlagConfig presets
    println!("\n--- 2. RedFlagConfig Presets ---\n");

    let strict = RedFlagConfig::strict();
    let relaxed = RedFlagConfig::relaxed();

    println!("  Strict config:");
    println!("    max_response_tokens: {}", strict.max_response_tokens);
    println!("    require_exact_format: {}", strict.require_exact_format);
    println!("    flag_self_correction: {}", strict.flag_self_correction);
    println!(
        "    confusion_patterns: {} patterns",
        strict.confusion_patterns.len()
    );
    println!("    min_response_length: {}", strict.min_response_length);
    println!(
        "    max_empty_line_ratio: {:.1}",
        strict.max_empty_line_ratio
    );

    println!("  Relaxed config:");
    println!("    max_response_tokens: {}", relaxed.max_response_tokens);
    println!("    require_exact_format: {}", relaxed.require_exact_format);
    println!("    flag_self_correction: {}", relaxed.flag_self_correction);
    println!(
        "    confusion_patterns: {} patterns",
        relaxed.confusion_patterns.len()
    );
    println!("    min_response_length: {}", relaxed.min_response_length);
    println!(
        "    max_empty_line_ratio: {:.1}",
        relaxed.max_empty_line_ratio
    );

    // 3. Validate responses with the strict validator
    println!("\n--- 3. Red-Flag Validation (Strict) ---\n");

    let validator = StandardRedFlagValidator::strict();

    for (i, resp) in responses.iter().enumerate() {
        let result = validator.validate(&resp.raw_response, &resp.metadata);
        let status = if result.is_valid() {
            "VALID".to_string()
        } else {
            format!(
                "FLAGGED: {}",
                match &result {
                    brainwires_mdap::red_flags::RedFlagResult::Flagged { reason, severity } => {
                        format!("{} (severity: {:.2})", reason, severity)
                    }
                    _ => unreachable!(),
                }
            )
        };
        println!("  Response {}: {}", i + 1, status);
    }

    // 4. Validate with relaxed config
    println!("\n--- 4. Red-Flag Validation (Relaxed) ---\n");

    let relaxed_validator = StandardRedFlagValidator::new(RedFlagConfig::relaxed(), None);

    for (i, resp) in responses.iter().enumerate() {
        let result = relaxed_validator.validate(&resp.raw_response, &resp.metadata);
        let status = if result.is_valid() {
            "VALID"
        } else {
            "FLAGGED"
        };
        println!("  Response {}: {}", i + 1, status);
    }

    // 5. Format-aware validation
    println!("\n--- 5. Format-Aware Validation ---\n");

    let json_validator = StandardRedFlagValidator::with_format(OutputFormat::Json);
    let one_of_validator = StandardRedFlagValidator::with_format(OutputFormat::OneOf(vec![
        "Paris".to_string(),
        "London".to_string(),
        "Berlin".to_string(),
    ]));

    let json_response = r#"{"answer": "Paris"}"#;
    let plain_response = "Paris";

    let json_meta = ResponseMetadata {
        token_count: 20,
        response_time_ms: 100,
        format_valid: true,
        finish_reason: Some("stop".to_string()),
        model: None,
    };

    println!(
        "  JSON validator on '{}': {}",
        json_response,
        if json_validator
            .validate(json_response, &json_meta)
            .is_valid()
        {
            "VALID"
        } else {
            "FLAGGED"
        },
    );
    println!(
        "  JSON validator on '{}': {}",
        plain_response,
        if json_validator
            .validate(plain_response, &json_meta)
            .is_valid()
        {
            "VALID"
        } else {
            "FLAGGED"
        },
    );
    println!(
        "  OneOf validator on '{}': {}",
        plain_response,
        if one_of_validator
            .validate(plain_response, &json_meta)
            .is_valid()
        {
            "VALID"
        } else {
            "FLAGGED"
        },
    );
    println!(
        "  OneOf validator on 'Tokyo': {}",
        if one_of_validator.validate("Tokyo", &json_meta).is_valid() {
            "VALID"
        } else {
            "FLAGGED"
        },
    );

    // 6. Show VoteResult structure
    println!("\n--- 6. VoteResult Structure ---\n");

    let mut distribution = HashMap::new();
    distribution.insert("Paris".to_string(), 4);
    distribution.insert("Lyon".to_string(), 1);

    let vote_result: VoteResult<String> = VoteResult {
        winner: "Paris".to_string(),
        winner_votes: 4,
        total_votes: 5,
        total_samples: 7,
        red_flagged_count: 2,
        vote_distribution: distribution,
        confidence: 0.80,
        red_flag_reasons: vec![
            "Response too long: 900 tokens > 750 limit".to_string(),
            "Self-correction detected: 'Wait,'".to_string(),
        ],
        early_stopped: false,
        weighted_confidence: Some(0.85),
        voting_method: VotingMethod::FirstToAheadByK,
    };

    println!("  Winner: {:?}", vote_result.winner);
    println!(
        "  Winner votes: {}/{}",
        vote_result.winner_votes, vote_result.total_votes
    );
    println!(
        "  Total samples (incl. red-flagged): {}",
        vote_result.total_samples
    );
    println!("  Red-flagged: {}", vote_result.red_flagged_count);
    println!("  Confidence: {:.0}%", vote_result.confidence * 100.0);
    println!(
        "  Weighted confidence: {:.0}%",
        vote_result.weighted_confidence.unwrap_or(0.0) * 100.0
    );
    println!("  Early stopped: {}", vote_result.early_stopped);
    println!("  Vote distribution:");
    for (candidate, votes) in &vote_result.vote_distribution {
        println!("    {}: {} votes", candidate, votes);
    }
    println!("  Red-flag reasons:");
    for reason in &vote_result.red_flag_reasons {
        println!("    - {}", reason);
    }

    println!("\n=== Done ===");
}

File: deno/examples/agents/voting_consensus.ts
// Example: MAKER Voting Consensus Building Blocks
// Demonstrates creating sampled responses with metadata, configuring red-flag
// validation, and validating responses against reliability criteria.
// Run: deno run deno/examples/agents/voting_consensus.ts

import {
  type MdapRedFlagResult,
  type OutputFormat,
  relaxedRedFlagConfig,
  type ResponseMetadata,
  type SampledResponse,
  StandardRedFlagValidator,
  strictRedFlagConfig,
  type VoteResult,
  type VotingMethod,
} from "@rullama/agent";

async function main() {
  console.log("=== MAKER Voting Consensus Building Blocks ===\n");

  // 1. Create sampled responses with metadata
  console.log("--- 1. Creating Sampled Responses ---\n");

  const responses: SampledResponse<string>[] = [
    {
      value: "Paris",
      metadata: {
        tokenCount: 12,
        responseTimeMs: 150,
        formatValid: true,
        finishReason: "stop",
        model: "claude-sonnet",
      },
      rawResponse: "The capital of France is Paris.",
      confidence: 0.5,
    },
    {
      value: "Paris",
      metadata: {
        tokenCount: 8,
        responseTimeMs: 120,
        formatValid: true,
        finishReason: "stop",
        model: "claude-sonnet",
      },
      rawResponse: "Paris",
      confidence: 0.95,
    },
    {
      value: "Wait, actually Lyon",
      metadata: {
        tokenCount: 45,
        responseTimeMs: 300,
        formatValid: true,
        finishReason: "stop",
      },
      rawResponse:
        "Wait, I think it might be Lyon. Actually, let me reconsider...",
      confidence: 0.5,
    },
    {
      value: "Paris",
      metadata: {
        tokenCount: 900,
        responseTimeMs: 2000,
        formatValid: true,
        finishReason: "length",
        model: "gpt-4o",
      },
      rawResponse: "A very long response that was truncated...",
      confidence: 0.5,
    },
  ];

  for (let i = 0; i < responses.length; i++) {
    const resp = responses[i];
    console.log(
      `  Response ${
        i + 1
      }: value="${resp.value}", tokens=${resp.metadata.tokenCount}, confidence=${
        resp.confidence.toFixed(2)
      }, model=${resp.metadata.model ?? "unknown"}`,
    );
  }

  // 2. Compare RedFlagConfig presets
  console.log("\n--- 2. RedFlagConfig Presets ---\n");

  const strict = strictRedFlagConfig();
  const relaxed = relaxedRedFlagConfig();

  console.log("  Strict config:");
  console.log(`    maxResponseTokens: ${strict.maxResponseTokens}`);
  console.log(`    requireExactFormat: ${strict.requireExactFormat}`);
  console.log(`    flagSelfCorrection: ${strict.flagSelfCorrection}`);
  console.log(
    `    confusionPatterns: ${strict.confusionPatterns.length} patterns`,
  );
  console.log(`    minResponseLength: ${strict.minResponseLength}`);
  console.log(`    maxEmptyLineRatio: ${strict.maxEmptyLineRatio.toFixed(1)}`);

  console.log("  Relaxed config:");
  console.log(`    maxResponseTokens: ${relaxed.maxResponseTokens}`);
  console.log(`    requireExactFormat: ${relaxed.requireExactFormat}`);
  console.log(`    flagSelfCorrection: ${relaxed.flagSelfCorrection}`);
  console.log(
    `    confusionPatterns: ${relaxed.confusionPatterns.length} patterns`,
  );
  console.log(`    minResponseLength: ${relaxed.minResponseLength}`);
  console.log(`    maxEmptyLineRatio: ${relaxed.maxEmptyLineRatio.toFixed(1)}`);

  // 3. Validate responses with the strict validator
  console.log("\n--- 3. Red-Flag Validation (Strict) ---\n");

  const validator = StandardRedFlagValidator.strict();

  for (let i = 0; i < responses.length; i++) {
    const resp = responses[i];
    const result: MdapRedFlagResult = validator.validate(
      resp.rawResponse,
      resp.metadata,
    );
    const status = result.valid
      ? "VALID"
      : `FLAGGED: ${result.reason.kind} (severity: ${
        result.severity.toFixed(2)
      })`;
    console.log(`  Response ${i + 1}: ${status}`);
  }

  // 4. Validate with relaxed config
  console.log("\n--- 4. Red-Flag Validation (Relaxed) ---\n");

  const relaxedValidator = new StandardRedFlagValidator(
    relaxedRedFlagConfig(),
  );

  for (let i = 0; i < responses.length; i++) {
    const resp = responses[i];
    const result = relaxedValidator.validate(resp.rawResponse, resp.metadata);
    const status = result.valid ? "VALID" : "FLAGGED";
    console.log(`  Response ${i + 1}: ${status}`);
  }

  // 5. Format-aware validation
  console.log("\n--- 5. Format-Aware Validation ---\n");

  const jsonFormat: OutputFormat = { kind: "json" };
  const oneOfFormat: OutputFormat = {
    kind: "one_of",
    options: ["Paris", "London", "Berlin"],
  };

  const jsonValidator = StandardRedFlagValidator.withFormat(jsonFormat);
  const oneOfValidator = StandardRedFlagValidator.withFormat(oneOfFormat);

  const jsonResponse = '{"answer": "Paris"}';
  const plainResponse = "Paris";

  const jsonMeta: ResponseMetadata = {
    tokenCount: 20,
    responseTimeMs: 100,
    formatValid: true,
    finishReason: "stop",
  };

  console.log(
    `  JSON validator on '${jsonResponse}': ${
      jsonValidator.validate(jsonResponse, jsonMeta).valid ? "VALID" : "FLAGGED"
    }`,
  );
  console.log(
    `  JSON validator on '${plainResponse}': ${
      jsonValidator.validate(plainResponse, jsonMeta).valid
        ? "VALID"
        : "FLAGGED"
    }`,
  );
  console.log(
    `  OneOf validator on '${plainResponse}': ${
      oneOfValidator.validate(plainResponse, jsonMeta).valid
        ? "VALID"
        : "FLAGGED"
    }`,
  );
  console.log(
    `  OneOf validator on 'Tokyo': ${
      oneOfValidator.validate("Tokyo", jsonMeta).valid ? "VALID" : "FLAGGED"
    }`,
  );

  // 6. Show VoteResult structure
  console.log("\n--- 6. VoteResult Structure ---\n");

  const voteResult: VoteResult<string> = {
    winner: "Paris",
    winnerVotes: 4,
    totalVotes: 5,
    totalSamples: 7,
    redFlaggedCount: 2,
    voteDistribution: { Paris: 4, Lyon: 1 },
    confidence: 0.80,
    redFlagReasons: [
      "Response too long: 900 tokens > 750 limit",
      "Self-correction detected: 'Wait,'",
    ],
    earlyStopped: false,
    weightedConfidence: 0.85,
    votingMethod: "first_to_ahead_by_k",
  };

  console.log(`  Winner: "${voteResult.winner}"`);
  console.log(
    `  Winner votes: ${voteResult.winnerVotes}/${voteResult.totalVotes}`,
  );
  console.log(
    `  Total samples (incl. red-flagged): ${voteResult.totalSamples}`,
  );
  console.log(`  Red-flagged: ${voteResult.redFlaggedCount}`);
  console.log(`  Confidence: ${(voteResult.confidence * 100).toFixed(0)}%`);
  console.log(
    `  Weighted confidence: ${
      ((voteResult.weightedConfidence ?? 0) * 100).toFixed(0)
    }%`,
  );
  console.log(`  Early stopped: ${voteResult.earlyStopped}`);
  console.log("  Vote distribution:");
  for (
    const [candidate, votes] of Object.entries(voteResult.voteDistribution)
  ) {
    console.log(`    ${candidate}: ${votes} votes`);
  }
  console.log("  Red-flag reasons:");
  for (const reason of voteResult.redFlagReasons) {
    console.log(`    - ${reason}`);
  }

  console.log("\n=== Done ===");
}

await main();
